### ISSUE TITLE
Will the UI elements look same on different platforms?

### ISSUE THREAD

**Issue by DaZiYuan** [dotnet/maui#50]
Does MAUI look different on different platforms, just like xamarin.forms?
I wish MAUI can keep the same UI on all platforms.

**Comment by charlesroddie** [comment]
Users mostly want identical UI. The ability to make breaking changes now would allow defaults for standard controls that appear as close as possible on different platforms.

Taking a dependency on SkiaSharp would allow some controls to have a single renderer which would automatically display identically across platforms, and improve the maintainability of the XF repo. Image-related controls including Icons, Images, Shapes would be particularly good candidates.

**Comment by JohnH-FF** [comment]
I have the opposite view, I want a droid button to look like a droid button on droid devices, and an iOS button to look like an iOS button on iOS devices. Users prefer a native experience, and a native look and feel.
It is not a common use case that users will have and use both devices at the same time. They do not need to look exactly the same. The layout can (and should!) be the same of course.

**Comment by Mrxx99** [comment]
It sounds like the best solution would be to support both and the  developer can choose

**Comment by JohnH-FF** [comment]
I would prefer they fix existing issues rather than split development efforts. I am all for MAUI, as long as they resolve the basic, fundamental problems that XF has. This is a good time to introduce breaking changes to course correct the platform.

**Comment by Depechie** [comment]
Just for reference @JohnH-FF a lot of companies want their 'branding' be known everywhere, hence the same look and feel on each platform ( look at the Facebook app or other major brands ). So this is actually a high valued request to have easy same design parity on each platform.
And yes, I agree the other side should also be available, having native look and feel per platform.

@Clancey showed his SKIA proposal yesterday during the MAUI overview Twitch stream, so if they decide to add it in the main repo, this would be awesome!

**Comment by JohnH-FF** [comment]
@Depechie Branding, absolutely. Icons, fonts, colours. Buttons though? What a user expects and what a company think they want are not always the same.

**Comment by dhindrik** [comment]
I like it when it keeps the native UI look. And if you want identical UI you can achieve it with Visual. Visual will probably be part of MAUI and it will be good if more visuals are added. But Visual is also a great way for third-parties to add "themes" to Xamarin.Forms/MAUI.

**Comment by GalaxiaGuy** [comment]
I realise this is descending into sharing anecdotes, but here is my experience:

* Just myself, I would prefer a native look and feel. I personally know a lot people (who tend to also be into technology) who feel the same way.
* Designers I work with tend to prefer consistency.
* I work for an agency, and our clients either want consistency, or don't care (i.e. they want the easiest and/or cheapest).
* Most users I know who aren't into technology, don't care and barely notice when an app is doing something more fitting for a different platform.

As a developer paid to build apps for others, I'd have to lean towards consistent cross platform controls.

**Comment by aguywithcode** [comment]
It looks like MAUI is bring back [Templated Controls ](https://github.com/dotnet/maui/blob/build/System.Maui.Core/IControlTemplated.cs) and raw visual elements. From WPF. I'm still browsing around but I saw a reference to Themes as well so best of both worlds...custom look and feel for those who want it. Platform specific themes who want that.

**Comment by TheRusstler** [comment]
I hope to make a real-world case for the "identical ui on all platforms" view:

- My experience in the business world is a strong customer desire for apps to look identical on all platforms.
- Our customers wish to choose hardware purely on it's physical merits, not its influence over the the app's design and UI controls.
- Our designers want to draw a beautiful creation, imagining everything down to the pixel, not "make it look and feel like every other [android|windows|ios] app out there".
- A consistent UI has huge benefits in terms of training - if a person is experienced with an app on one platform, they would like (and expect) it to look and feel identical on another platform when switching for personal, or company-imposed change.
- This view applies not just to business apps, but in the consumer world as well. As @Depechie mentioned, apps like facebook, whatsapp, messenger etc. all try to make the experience the same across platforms for simplicity and to conform to their branding and design.

Historically we have lived in a divide of Apple vs Google vs Microsoft user loyalty, whereby individuals want all apps to follow the specific style of their preferred company.

I feel the future lies not in greater divide, but in a focus on designing the best user experience as the designer imagined it, pixel-perfect down to the user's preferred physical device.

**Comment by GiampaoloGabba** [comment]
Maybe i'm missing something, but now i'm already able to get the same look and feel on ios and android just using styles (with or without material theme) & some platform effect.

I like to have a general look and feel of the app wich is almost the same, then there are the "functional" details wich are specific to the platform: like the app bar if i choose the native one, the native dialogs & pickers. But thats fine because that are small functional parts  that use the OS pradigms that users are able to.

Btw with all the new things that are coming (appbar, tabview, path, shapes, shadows) the process to streamline the look and feel will be easier.

Moreover, with the new acritecture it will be much easier to create and integrate custom control with skia!

**Comment by legistek** [comment]
> It looks like MAUI is bring back [Templated Controls ](https://github.com/dotnet/maui/blob/build/System.Maui.Core/IControlTemplated.cs) and raw visual elements. From WPF. I'm still browsing around but I saw a reference to Themes as well so best of both worlds...custom look and feel for those who want it. Platform specific themes who want that.

Don't read too much into that. So far MAUI is almost a complete copy of Xamarin Forms and `IControlTemplated` has long (always?) been a part of XF but isn't for templated controls the way we think of them in WPF.

That said I agree I would like to have the option of lookless/templated controls.

**Comment by BillHenning** [comment]
Coming from a WPF/Silverlight/UWP background, it would be very valuable to support WPF-like ControlTemplates along with content DataTemplates.  Those mechanisms make it so easy to customize UI rendering.

**Comment by bkaankose** [comment]
I want to look at this discrimination from different angle.

The root reason why business wants every page/control to look same is they are not aware of design guidelines for each platform and they think having 1 consistent UI will cost them a lot less than adjusting/supporting every visual element for different platforms. This is wrong and it is not Xamarin's or MAUI's job to fix.

Coming from Xamarin and UWP background, I personally prefer to have native feel for each platform. However, I also expect that button I put on the view must be treated well by the framework (Xamarin or MAUI in this case) that I should not create extra margins, different renderer-like classes to support to achieve this layout issues. I put it there, it is native, it's not identical between platforms but it is similar as highly it can get in terms of layout, it is not rendered by Skia, it has all the effects/shadows that native platform provides, it respects the OS version I am in and it does position itself well with other controls or UI containers I put into. This may seem a lot, but this is something I observe from the Xamarin's history. They tried to achieve this and came a long way. Thrust me, you were not able to just create a single basic page and make it perfectly layout in different platforms 4-5 years ago. 

Personally, their biggest mistake was to support both stupid business ideas and develop new features as the framework develops. They forced themselves too much behind this 'supporting business by making platform highly customizable in order to lower development effort' idea, that eventually made the platform lose in the competition. And this lose was caused by the nature of Xamarin Forms. This nature relies highly on the platform specific and nature feeling of controls. You can't deny this fact of the framework by saying "Here you go all the redundant stuff that is highly against our platform's core nature we wasted team effort on instead of strengthening our framework core that can be super efficient for the developers that know what they are doing". :) 

Other platforms that powered by graphics libraries are not good candidates for this approach. Because they intended to be "Same feeling on all platforms" approach. But Xamarin or MAUI is not, and should not because it is not an efficient way to support 'design and develop' approach.

TDLR; the problem is not the framework itself, it is the business and no framework should change it's core for the sake of supporting stupid business ideas...

**Comment by legistek** [comment]
> TDLR; the problem is not the framework itself, it is the business and no framework should change it's core for the sake of supporting stupid business ideas...

LOL. The essence of what's wrong with so many developers. You think the business and the users should adapt to you and not the other way around.

Maybe you haven't noticed but the overwhelming majority of apps are WEB FIRST now and when they do have native versions they go out of their way to make the native version look and feels just like the web version. This is thanks to React Native and Ionic, etc., that allow you to use the same markup and code in every platform. That's why React Native and Ionic have been dominating while XF - not to mention any native app platform including Android/Java and iOS/Swift - have been falling behind HTML/Javascript.

Businesses WANT uniformity for brand recognition, consistency, and user familiarity. You get the same experience from the application no matter what device you're on. That's what modern users expect. 

Native apps in general are also dying. The only native platforms I actually care about are Windows and MacOS and I want my app to look identical between them. I'd like to be able to use the same codebase for web too but if I have to have a separate one, then I'll make it mobile friendly and publish it to iOS and Android as a PWA or use React Native. 

Bottom line is my business users couldn't care less about having a "native" iOS or Android app that looks like a "native" iOS or Android app. They don't know the difference between native and PWA and don't care whether it uses Apple chrome or Android blue as long as they can do what they need it to do.

**Comment by bkaankose** [comment]
@legistek My point is "what you are saying is against Xamarin's nature and it should not keep working against it." This is what exactly you want from MAUI.

MAUI should not do the same mistakes that Xamarin did (even though it looks like they share a lot as today) just because Ionic or React Native achieved this popularity by providing WEB technologies in mobile apps. Web is web, mobile is mobile. If you want all your app to feel the same, change your web to support mobile resolutions and create PWA. I don't want to get involved in the discussion on how PWAs or any web supported native-like bundles on mobile devices killed the having properly designed/developed apps. If you are a web developer and want to create mobile app, Xamarin is not for you, even though the team is trying to make it for you for the last 3 years. If you are a mobile developer that knows a little bit XAML and has knowledge about platform specific features that Xamarin core provides you, then use Xamarin to achieve your native app. By native I don't mean the outcome of your code or development only but in terms of respecting OS guidelines.

What is the point of having design guidelines for each platform if we want to have same view all the time? What is the point of even creating a native app if we can use web for everything? This 'let's create everything in web because it's cost effective and easy' approach is completely wrong in my opinion. Xamarin did it for years and lost. Let's hope MAUI learns from it.

**Comment by JohnH-FF** [comment]
> LOL. The essence of what's wrong with so many developers. You think the business and the users should adapt to you and not the other way around.

That isn't what we are saying at all. Business needs to understand users, and users (mostly) want a native experience. They do not care about how it looks on iOS if they are using Droid.

**Comment by Depechie** [comment]
Only thing that I care about is how much do I need to fight with the framework/tools to get paid. And for all the apps that I created that meant design parity on all platforms and not having the original native look. So in that sense @bkaankose is right, maybe I just need to pick another tool ( web or flutter ). But since I love C# I would rather stick with MAUI.

**Comment by legistek** [comment]
> and users (mostly) want a native experience. 
That's not my experience. In my experience they don't care. And clearly Microsoft disagrees. Look at Office for Windows vs. Mac vs. iOS vs. Android. They're amazingly consistent. I frankly wish they would just open up to us whatever platform they're using to achieve that rather than go through all this trouble of making something new.

> They do not care about how it looks on iOS if they are using Droid.
True but iOS vs. Android is the wrong comparison. Desktop Browser vs. mobile browser/PWA vs. mobile native is what's relevant. If they use a web app on their desktop/laptop, and then you make them a version for their mobile device - whether it's PWA or native - they're going to expect it looks the same as the web version. After all, the web version will look the same no matter the device, and since the mobile version needs to look like the web version, the mobile version also needs to look the same regardless of device.

I also don't give two hoots what designers in the Apple Circle or the Googleplex think an app should look like. I care what *my* designers think *my* app should look like and that it looks beautiful and consistent on all the devices it will be used on. 

The only way to make an argument for the old XF approach of having controls look native on each device is to believe that native mobile is still a relevant platform unto itself. It's not (other than for games - but games almost always use identical UIs across devices). The user gets little-to-no additional value from a native mobile app written in Java or Swift versus an app written in React Native or Ionic. The business, on the other hand, gets tremendous value from having a unified design and codebase be realized on multiple platforms simultaneously. This is why React and Ionic are dominating. MAUI needs to follow that lead and do for C#/XAML what those platforms did for HTML+Javascript, or it will be DOA.

**Comment by Clancey** [comment]
Both sides are correct. Some devs/users need native look and feel. And some need a single look/feel regardless of the OS. And some people want to mix and match at the control level.  

To start Maui will look and feel like Xamarin Forms witch gives you native look and feel. But we will continue to improve upon the identical look and feel.

**Comment by saint4eva** [comment]
> I would prefer they fix existing issues rather than split development efforts. I am all for MAUI, as long as they resolve the basic, fundamental problems that XF has. This is a good time to introduce breaking changes to course correct the platform.

Could you name some of the fundamental problems? So the team would know issues to tackle?

**Comment by legistek** [comment]
@Clancey true that there might be need out there for both, but native look can always be achieved from a lookless framework. The inverse is not true.

My biggest technical issue with the XF rendering system is that it's bloated. Every platform shouldn't need to implement dozens of renderers. If a particular platform offers a special control and it makes sense to implement it natively then that can be an added bonus, but there should be a lookless fallback "default implementation" for everything that is built on primitives. 

Alternatively, perhaps a "Presentation" branch of MAUI is called for. `System.MAUI.Presentation`. It would be mapped to the WPF/UWP namespace for compatibility (`http://schemas.microsoft.com/winfx/2006/xaml/presentation`).
Controls under this branch would all be lookless and rely on renderers that implemented only a small number primitives. At the same time you could keep the WPF/UWP naming convention for everything in this namespace.,  

I actually won't be upset if the official project doesn't go this route, as long as certain things in XF are corrected which prevents this now. XML namespace mapping is a big one and also the overzealous use of `internal` and `sealed`. I could make a PR that would address these and allow someone like myself to experiment in this direction without Microsoft having to commit to an all lookless UI out of the gate.

**Comment by XeonG** [comment]
JavaFx looks the same on all platforms.. same with QT, I like both visually.. why? don't care about the native UI platforms crappy design decisions and lame implementations of various UI elements and how they work..... or not properly....it also means adding better support in functionality or improving styling can be done once and work across all platforms more easily.

It's also why some people make apps in game engines like Unity.... the GUI and UX can work the same across all platforms. 

> I also don't give two hoots what designers in the Apple Circle or the Googleplex think an app should look like. I care what my designers think my app should look like and that it looks beautiful and consistent on all the devices it will be used on.

Yup.

**Comment by legistek** [comment]
Also if we're being honest, let's remember that when Xamarin started, iPhones and Android devices were weaklings compared to now. I don't know if anyone ever tried this but my strong suspicion is that they weren't nearly powerful enough for the application layer to build and manage the entire visual tree down to every glyph, line, and brush, not to mention bubbling all the low level input events, translating screen coordinates, etc.  Maybe the "native look on each platform" idea had some independent appeal, but I suspect there was no other choice and so it HAD to be spun as a good thing.

Things are much different now and there's no question even the cheapest iPhone could handle a lookless UI framework. So let's at least recognize that the major _technical_ argument for native look that did exist in 2011 no longer applies.

**Comment by AndrewPlakhotnyi** [comment]
OS specific UI/UX it is not just about how controls look. It is also another layout, interaction patterns, animation, dialogs and even logic.

Changing the style of a button to a native one won't bring your app a native look. You must change the whole application to give it a truly native feel.

**Comment by JohnH-FF** [comment]
@legistek 

> > They do not care about how it looks on iOS if they are using Droid.
> 
> True but iOS vs. Android is the wrong comparison. Desktop Browser vs. mobile browser/PWA vs. mobile native is what's relevant.

Both are relevant when the argument is to be exactly the same across all platforms.

> If they use a web app on their desktop/laptop, and then you make them a version for their mobile device - whether it's PWA or native - they're going to expect it looks the same as the web version. 

Umm, no they don't. They expect the layout to be different, complex data entry screens on a desktop app are ok, but not on a mobile app. That is why we have separate layouts for iPhone and iPad to cope with the different form factors. You want screens optimised for each platform. On mobile that complex data entry screen might be broken into several pages, or a wizard style.

> I also don't give two hoots what designers in the Apple Circle or the Googleplex think an app should look like. I care what _my_ designers think _my_ app should look like and that it looks beautiful and consistent on all the devices it will be used on.

That is your choice if you want to ignore the style guides on each native platform. The style guides that help apps become easier to use for users that use that platform every day.

**Comment by legistek** [comment]
> Umm, no they don't. They expect the layout to be different, complex data entry screens on a desktop app are ok, but not on a mobile app. That is why we have separate layouts for iPhone and iPad to cope with the different form factors. You want screens optimised for each platform. On mobile that complex data entry screen might be broken into several pages, or a wizard style.

We're not talking about desktop v. mobile _layout_ we're talking about styling.

> That is your choice if you want to ignore the style guides on each native platform. The style guides that help apps become easier to use for users that use that platform every day.

Yes I do want to ignore them. And so do Microsoft, Facebook, Twitter, Netflix, Instagram, Spotify, Slack, Uber, Grubhub, Youtube, and Amazon. Need I go on? Are they all beholden to "stupid business ideas"? Or maybe they have actually done the market research and realized uniformity is the way to go.

**Comment by kevinjpetersen** [comment]
I have to agree with @legistek about having the option to style our app the way WE want to style our app, and not a way that is dictated by certain guidelines or because some people "say that's the way it's done".

@JohnH-FF I kindly disagree with your argument that you keep bringing up, that you say that USERS do and don't want X and Y. No one knows what the users want, final. We have an idea of what might make them happy, but we can't speak on behalf of them, that's just silly in my opinion.

When apps are developed, designers and developers make their OWN decision what to make. It's THEIR app, and that shouldn't be constrained by silly guidelines just because some blog post or someone once said "that's how it's done, I know better so listen to me".

I understand that people might have done this in the past, the majority, however we're in a world now where everyone has different needs, everyone has different opinons, perspectives, ideas, you get the point.

So overall, no I don't think Native look only is valid anymore. It might have been for a while, but even for a while aswell, it haven't (looking at the examples that a couple of people have brought up like Facebook, Messenger, Discord, Twitter etc.)

**Comment by GiampaoloGabba** [comment]
well, we already have the option to style our app as we want. I do it all the time. My apps are 99% the same in ios and android. 
I think we should discuss here how we can improve this beahviour, make it smooth and easier and give us more freedom.

Specs like shapes, gradients, shadow, appbar, tabview are all great and heading to

**Comment by JohnH-FF** [comment]
@kevinjpetersen I never said listen to me, I pointed out that there are style guides for each platform. I also said feel free to ignore them, but I wont be.
Regarding users, do you believe they want every app to look completely different to each other? That they want a steeper learning curve because an app designer thinks it would be better for them? If so, fair enough.
Everyone is entitled to their opinion, and so am I. Respectfully, I disagree. If the new slim renderer process resolves that for us all, then great. I will still use native controls on each platform, styled the way we want, but native nonetheless.

**Comment by kevinjpetersen** [comment]
@JohnH-FF In terms of if users want every app to look completely different to each other, I think that's up to the user themself and their preference. I don't think you can just say that the majority of users want a native look and feel, but ofcourse you can't say the opposite either that they want a non-native look and feel.

In the end, designers and developers make the app. Taking Facebook and Discord as an example, they basically don't have any native look to them at all, and alot of people like that, and alot of people don't like that. It's all in all a opinion based area where I don't want to see people being forced to do X if you want to do Y.

**Comment by GiampaoloGabba** [comment]
Well, we already have the option to style our app as we want. I do it all the time. My apps are 99% the same in ios and android. Its not hard. I dont even use visual but a couple of effects, but with Vissual is even simplier.
This should be clear: it is already possibile to achieve the same style on both platform (with minor differences), otherways new people approaching this project will get incomplete informations from the start :)

I think we should discuss here how we can improve this behaviour, make it smooth and easier and give us more freedom.

Specs like shapes, gradients, shadow, appbar, tabview are all great and heading to address this.

I dont like to have a complete custom drawing canvas. One of the best things of Xamarin is being able to deeply integrate our app/views with the native platform and i love it, the possibilities are endless!

But i repeat, we need for sure more freedom on how to do our layouts (always with maximum performance and optimization possibile).

**Comment by kevinjpetersen** [comment]
@GiampaoloGabba Agreed, right now in XF we CAN get our apps to be similar rather quickly, altho I would say that the way that's done could be improved alot.


I think the main discussion of this thread has been derailed, and I for once would like to get back on topic. I think the whole point of this thread is to discuss how to make it even more simpler to have the exact same UI for all platforms.

If you want to discuss native look and feel, then I don't think that fits in here as it's a thread for UI looking the exact same. Having a seperate thread for native look would make sense.

**Comment by DaZiYuan** [comment]
From a developer's perspective, when you create an apple, you want it to remain an apple rather than looking like a pear in the Android world or orange in IOS. Unless it is of your own will, You can customize it to work that way.
Don't talk about users anymore, you're the creator, and you just want apple to work your way no matter in what platform.

So I like the way of flutter, But I prefer to write it in XAML. That's why I asked the question

**Comment by bkaankose** [comment]
Any style you apply to customize your controls in Xamarin will be applied to View and View can not be performant in Xamarin because it is not how Xamarin works. In this case for MAUI as well. Since any property you change must have the corresponding native binding to make it look native, all the custom stuff is handled by View. This is why we have renderers. In order to make your control to look same in all platforms, whole control must be a huge View and that's why Xamarin is slow. View is slow, but renderers are fast. This is the key point that is only special to Xamarin.

You don't want auto-capital labels on your Android button? Go extend your renderer and disable it natively. Do not create a custom View that acts/feels like a button because it will kill all your native performance. Then people started to hate renderers because for them it requires a lot of effort to do small things. I can't deny that, but it is the most efficient way to achieve such stuff because of how Xamarin works.

I feel like we're getting out of the topic on this thread. Therefore I will finish my thoughts by saying 'Whatever we do to improve this framework and get more popularity in the industry, we must keep in mind how Xamarin is intended to work and not force MAUI to become more available/less performant version of Xamarin.'

**Comment by kevinjpetersen** [comment]
@bkaankose That last statement of yours I have a hard time agreeing with. Saying that we must keep in mind how Xamarin is "intended" to work is old thinking in my opinion. Xamarin is intended for you to do ANYTHING you want to do.

Now is it more optimized towards the native side? Obviously. But does this mean it shouldn't focus on Same UI across all Platforms? No, it doesn't. :)

MAUI is basically Xamarin, correct, but I don't think you should directly compare it to that. MAUI is meant as the new Xamarin, so the "what's intended" doesn't apply in my opinion.

**Comment by legistek** [comment]
So there is an old open Xamarin Forms issue (https://github.com/xamarin/Xamarin.Forms/issues/2452) that was directly on point to this. It was initiated by one of the Xamarin team so I think many of us hoped that it would actually happen. Unfortunately it seems dead and the discussion veered off into Skia advocacy (again I don't understand why that's even relevant at the app layer) and hating on XAML.

Point is if we had drawing primitives like they were talking about adding, and efficient renderers then we or third parties could independently make lookless controls that are uniform on each platform without diminishing anyone's ability to surrender their design choices to Apple if they want to.

That's really not fully possible now. It's _somewhat_ possible but there are a ton of limitations particularly when you want to affect more than just static look but also input response, animation, etc. And if you want to do anything with touch you're completely at the mercy of the native implementation now. Further, as I mentioned earlier, with all the `internal`s and `sealed` classes in XF we'd still have to fork to get it to where it would need to be. I know because I've tried. 

That said, it's all doable. Someone just needs to lead the charge. I find it ironic that Microsoft isn't doing so given that their flagship application platform - Office - is a model for what so many of us are screaming for. Web, Desktop, Mobile, all looking and behaving nearly identically. I get if they don't want to give away ALL their proprietary toys, but at least give us a fighting chance. 

Bottom line, if anyone expects "MAUI" or whatever it winds up being called to seriously compete with React Native and Ionic, this isn't even a question. On the other hand if Microsoft is content for XF to remain a niche platform with a new name for people who want to build simple mobile apps in C# and XAML and that's all it's useful for, then best of luck but I'll pass. That's not evolution that's living in the past.

**Comment by PureWeen** [comment]
All the planned features in the roadmap 
https://github.com/xamarin/Xamarin.Forms/wiki/Feature-Roadmap

Will let you take the route of looking the same on all the platforms. @jsuarezruiz even did a few spikes with his drawing implementations where he drew a slide control that looked the same on all platforms. We've talked about providing a Visual="XAML" (name pending) that would basically render the control using purely xplat features (like the new appbar and tabview) at which point you can influence the entire control template (like in UWP) 

The native controls aren't going to go away for those that want it, we'll still have material, we're looking at fluent, and we're pushing hard on the drawing (as you can see in the roadmap) so users can have full control.

I'd recommend taking a look at the specs in the roadmap and indicating where any of those spec fall short of your expectations. I also want to highlight that that roadmap is all pre-Maui.

> That's really not fully possible now. It's somewhat possible but there are a ton of limitations particularly when you want to affect more than just static look but also input response, animation, etc. And if you want to do anything with touch you're completely at the mercy of the native implementation now.

These are the areas we want Maui to shine, the slim renderer architecture we're proposing will fully open up the native implementations plus the ability for Maui to operate in a completely multi targeted environment will connect the developer to native implementations.

https://github.com/dotnet/maui/issues/28

If the slim renderers/multi targeting still fall short for you please let us know on those specs!!


We'll have to see where devs take things but honestly I think AppBar/TabView/Xplat Shell/XAML rendered controls (not skia)/All the Drawing stuff @jsuarezruiz has

Will mostly take over. But the idea is you'll be able to choose. Some places you'll keep the native tabs and other's you'll use our XAML ones.

**Comment by charlesroddie** [comment]
> @PureWeen Visual="XAML" (name pending) ... XAML rendered controls

I'm trying to understand this post and I assume this means Visual rendered controls. It seems to make sense like that. XAML refers to the markup language. Now MS is using XAML to refer also to UWP (UI.Xaml) whether or not it uses XAML the markup language. It makes a large proportion of issues in that repo impossible to understand. Now it also refers to Xamarin.Forms.Visual? Please don't do that.

**Comment by PureWeen** [comment]
Honestly the Visual was going to be called "Forms" but now we're called Maui so I don't know what to call it. That's not super relevant to the point. 

Let's just call it "thing to be defined"

Basically it's just constructing a Button/Slider/etc.. out of Maui elements 

So a grid, a label, a touch recognizer

All things that you can customize. For example the appbar/tabbar are all just constructed using grids/labels/etc... so anyone can modify those how much ever they want to

**Comment by legistek** [comment]
> Honestly the Visual was going to be called "Forms" but now we're called Maui so I don't know what to call it. That's not super relevant to the point.

@PureWeen why not `Control`? Or `TemplatedControl` That's what we're talking about right?

Anyway thanks a lot for jumping in on this topic. @Clancey's comment earlier in the thread seemed to suggest that uniform look wasn't going to be a focus of the first release. But I'm glad to hear it will be.

**Comment by kevinjpetersen** [comment]
@PureWeen Just to get this straight, that means that MAUI Elements in combination with the "thing to be defined" can be used to create a UI that looks the same for all platforms? Similar to how Visual=Material does it currently in XF?

**Comment by kevinjpetersen** [comment]
@PureWeen .. and by Similar to Visual=Material, I mean not having to write any custom controls, but just access a list of predefined controls like "Button, Slider, Checkbox, etc." and applying the Visual="Thing to be defined" ?

**Comment by PureWeen** [comment]
> Just to get this straight, that means that MAUI Elements in combination with the "thing to be defined" can be used to create a UI that looks the same for all platforms? Similar to how Visual=Material does it currently in XF?

Remove the word MAUI from your sentence and then yes!

 I want to clarify a bit that this isn't a Maui thing this is a by the end of the year as part of Forms thing. All of the features I'm talking about are on the XF roadmap https://github.com/xamarin/Xamarin.Forms/wiki/Feature-Roadmap

We're still figuring out the best way to expose this idea out. Whether it should be through a Visual or adding ControlsTemplates to everything.

Adding ControlTemplates to everything is a bit weird because then what about native controls? 

We have a spec here about ways we want to customize visuals
https://github.com/xamarin/Xamarin.Forms/issues/5005

A neat idea @jsuarezruiz had is that we could introduce a VisualTemplate that would be specific to the visual you specify.

So users could possibly do something like this

```XAML
<Button>
<VisualTemplate>
<ButtonControlTemplate>
<grid><label></grid>
</ButtonControlTemplate>
</VisualTemplate>
</Button>

```

**Comment by kevinjpetersen** [comment]
@PureWeen The downside and bad part of the example at the bottom (in my opinion) is that users have to write alot more code. I personally think if there's a way to make controls just a single liner, that would make the process of developing apps ALOT faster, while also being streamlined.

If you mean that syntax for making the Controls for the Visual= thing, then that could probably work yes, but for the users that just want to "Create Project" and then just start coding in XAML, they shouldn't have to type the above, simply to make controls look the same on each platform :)

IMO :)

**Comment by PureWeen** [comment]
@kevinjpetersen right!

All I'm articulating is how users would customize. Visual propagates through your entire hierarchy so you can set a Page to 

Page Visual="MakeItAllTheSamePlease"

At which point all your controls will just use the visual setup for that which renderers the controls using xplat elements

BUT I also want to emphasize this isn't something we're directly promising yet :-) If we're actually going to build all these controls.

We'll start with appbar/tabview and then probably start building out from there. 

And then once we have the mechanisms for building these controls we'll take any PRs from people that want to create the symmetrical xplat controls.

So the main thing I'm wanting to know from everyone. What are the limitations of these specs https://github.com/xamarin/Xamarin.Forms/wiki/Feature-Roadmap

Where do they fall short of achieving the points brought up on this issue?

**Comment by legistek** [comment]
> We're still figuring out the best way to expose this idea out. Whether it should be through a Visual or adding ControlsTemplates to everything. Adding ControlTemplates to everything is a bit weird because then what about native controls?

Agreed not everything would be derived from `Control`. That's why I was suggesting a whole new `System.Maui.Presentation` namespace mapped to `http://schemas.microsoft.com/winfx/2006/xaml/presentation` that uses the WPF/UWP naming conventions, for everything from `ContentControl` to `ItemsControl` to `ButtonBase`, etc. If you want to go ALL lookless, like I would, you say `xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation". If you want to go ALL native,  you say `xmlns="http://xamarin.com/schemas/2014/forms"`. If you want to mix and match, you can do so through ns prefixes. Either way there's a clear distinction between controls that will use native styling and those with uniform styling and you can pick and choose. `System.Maui.Button` would be drawn by the native renderer but `System.Maui.Presentation.Controls.Button` would be templated, with naming conflicts avoided via the ns's.

I'm not sure I see the value in a special template for each control? At minimum we would need to be able to make our own so the base for `ButtonControlTemplate` needs to not have internal constructors and such that prevent us from subclassing. As an avid WPF/UWP fan though I would of course urge you to stick with `ControlTemplate` qualified with a `TargetType` for validation purposes. 

> Where do they fall short of achieving the points brought up on this issue?

This (https://github.com/xamarin/Xamarin.Forms/issues/6452) is definitely a blocker right now because it prevents doing this "xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation".

There are also issues with `IControlTemplated` being internal and the way `TemplatedParent` binding works, which I tried to fix when I implemented `RelativeSource` binding but part of it had to be undone for compatibility. Essentially it's impossible now to use `ControlTemplate`s with anything other than a `ContentView`. If we just added `Control` to XF Core - with all the relevant templatable properties like `Background`, `Foreground`, etc., and had it implement `IControlTemplated` that would work. 

Finally input events, bubbling, and coordinate transformation need to be added in at the platform and framework level so a spec is needed for that. I assume we just want the platform to just do the bare compositing of primitives without necessarily having to be aware of the visual tree, and provide low level screen/mouse input events to the framework, letting the framework do the routing.

Otherwise I think it's just a matter of writing the renderers for the primitives and settling on what they will be.

All my contributions to the XF repo have really been with this eventual goal in mind so please I'm very willing and happy to help further to steer it this direction if that's where it's going.

**Comment by legistek** [comment]
> @PureWeen The downside and bad part of the example at the bottom (in my opinion) is that users have to write alot more code. I personally think if there's a way to make controls just a single liner, that would make the process of developing apps ALOT faster, while also being streamlined.
> 
> If you mean that the syntax is for making the actual Controls for the Visual= thing, then that could probably work yes, but for the users that just want to "Create Project" and then just start coding in XAML, they shouldn't have to type the above, simply to make controls look the same on each platform :)
> 
> IMO :)

I was assuming from the example that `ButtonControlTemplate` could be a `StaticResource` so you could just say:
`<Button VisualTemplate="{StaticResource MyButtonTemplate}"/>`

Furthermore I would assume/hope this could be incorporated into a `Style` which you could set a default for in a resource dictionary.

Basically I just hope/want it to be cross-platform WPF. Ok? I proudly admit it. :D

**Comment by PureWeen** [comment]
>  I was assuming from the example that ButtonControlTemplate could be a StaticResource so you could just say:
> <Button VisualTemplate="{StaticResource MyButtonTemplate}"/>
> 
> Furthermore I would assume/hope this could be incorporated into a Style which you could set a default for in a resource dictionary.

Yes

> Basically I just hope/want it to be cross-platform WPF. Ok? I proudly admit it. :D

Hehe.. I'm a really big fan of WPF control templates. Anytime I have to customize a control natively UWP is typically the least headache. 

I'm hoping we can bring a lot of that ease to the table with our current road map!!

**Comment by GiampaoloGabba** [comment]
There are a lot of useful informations in this thread, i love it :)
The only thing that worries me a bit are performance.
If we can customize all the control with an hypothetical Visual="Whatever", that  replaces all the native control with custom ones created from MAUI elements... well, potentially there may be many more native views added to the visual tree, resulting in a lot of layout measurements, wich are killing performance on mid/low-end android phones.
